### DISCUSSION TITLE
Welcome to snitch Discussions!

### DISCUSSION THREAD

**Discussion by cschreib** [snitch-org/snitch#36] [Announcements]
## 👋 Welcome!

The GitHub Discussions page is the right place to come and ask questions and share your experience on:
 * how to use _snatch_,
 * general testing approaches and best practices.

If you need to report a bug in _snatch_, or have a suggestion for a new feature, please search the [issue tracker](https://github.com/cschreib/snatch/issues) in case it has already been reported/suggested. If not, then please [open an issue](https://github.com/cschreib/snatch/issues/new). 

You can even contribute to _snatch_ if you feel like it, [Pull requests](https://github.com/cschreib/snatch/pulls) are always welcome! Just make sure you read the [roadmap](https://github.com/cschreib/snatch/blob/main/doc/comparison_catch2.md) first.

**Comment by e-dant** [comment]
Hi there --

First off, thanks for your work. Mostly just here for a big +1 to Snatch.

Less importantly, my 2C for the roadmap: I'd prefer as few features as possible. Do one thing, do it well and all that.

**Reply by cschreib** [reply]
Thank you for the kind words and the feedback! 

About the roadmap: this project has several conflicting goals. One is to be a lightweight test framework, and the other is to be user-friendly. For the latter, I want to support as many "good" features of _Catch2_ as possible, to make migration easy. But this is limited by the first goal, which will require drawing a line in the sand somewhere. The current roadmap is an attempt at a compromise between these two goals.

**Comment by willwray** [comment]
Hi, and thanks for polishing and publishing.

I wonder if you considered Boost license. I believe that the main difference vs. MIT is the level of copyright attribution (BSL still requires to preserve in-source notices IIUC). Catch2 uses it. It seems appropriate for a derivative project to follow the lead.

FYI I'm looking for a simple dependency like this as a replacement for boost.ut which proved too experimental for my taste. Looking forward to giving it a spin.

**Reply by cschreib** [reply]
Hi and thanks for the feedback! Is there anything in particular you dislike about the MIT license? It is my default OSS license because it is well known, small, simple, and very permissive. But I must admit I have not given it a lot of thoughts.

**Reply by willwray** [reply]
BSL is smaller, simpler and more permissive than MIT, and it _should_ be better known, so it may fit your project well.

FOSSA 10 May 2022 [Open Source Licenses 101: Boost Software License](https://fossa.com/blog/open-source-licenses-101-boost-software-license/#:~:text=The%20BSL%20and%20MIT%20Licenses,is%20in%20compiled%20code%20form)

We could ask Phil why he chose BSL for Catch (he posted responses in the original reddit announcement)

**Reply by cschreib** [reply]
Thank you for the link, that was instructive. I never realized the MIT license actually required attribution in binary form. I always interpreted "the software" as the source code... Consider my mind changed!

**Reply by willwray** [reply]
Great stuff. This should make the library more attractive, to those in the know at least.

**Comment by willwray** [comment]
Only me! Next stop, the name...

Phil famously regretted using 'catch'; a clever name that was hard to search on because "catch C++" hits a long list of links on exception handling. Hence Catch2.

Snatch is similarly clever, less likely to have issues with search, then is 'cheeky' and might bring negative connotations. It made me fnar. Then reflect.

Not just the rubbish homonymous Brit-flick, there's the American anatomical vulgarism and the implication of misappropriation.

Of course, I wouldn't be raising this exception if I didn't have an idea...

v1.0 could be occasion for a rename, and renaming a reason to re-announce.

**Reply by cschreib** [reply]
🤦 must every word in the dictionary be used to designate sexual organs? I am not a native speaker, I didn't know this meaning of the word. Looks like finding an alternative would be appropriate, then.

You said you had an idea; feel free to share it. Good names are hard to come by! I'll have a think in the meantime.

Thanks for raising this.

**Reply by willwray** [reply]
Yeah, naming is hard, and can be a minefield. It's important though.

The idea I hinted at is also a bit 'cheeky' in following Catch and Catch2:

  Try2 test

try comes before catch, or does it...
"try to" is aspirational - we all aim to test more than we actually do.

Adding the 2 even for version 1.0 makes it more searchable, and is also a nod to Catch2 which is now at V3 but keeps the name at 2 perpetually.

Also, if Phil complains, you have someone else to blame :wink:

**Reply by cschreib** [reply]
Ah that's clever! But I don't really want to follow the trend of having what looks like a version number in the project name (_Catch2_ isn't the only one with this problem, there is also _sol2_). It does help with searchability, but I think it also confuses people. It certainly confuses me :)

What about _snitch_? No anatomical vulgarisms I should know about?

**Reply by cschreib** [reply]
I kind of like _snitch_, as in it's a library that may have its own bugs and faults, but it tells you about some other program/library's faults.

**Reply by e-dant** [reply]
`segwork`

**Reply by willwray** [reply]
Even my filthy Finbarr Saunders mind can't see vulgarity in 'snitch', just the usual tawdry low-life images from crime thrillers.

<details><summary>At a stretch, since you ask;</summary> in a free-rhyming train of thought, its alliterative cadence brings to mind 'switch bitch', a book of short stories by Raold Dahl (absolutely not a children's book, but one of which I read at an inappropriately early age in a sci-fi anthology taken from my Grandma's book shelf).</details>

But I digress.

Snitch ain't bad, now you explain the logic; "snitch c++" should hit it in 1.

'Snitch' is also semantically related to 'nose', a name already related to test in a Python test automation framework. Sniffing out bugs and errors. The logo could then be something nose-related in order to reinforce this interpretation over the murky image of an informant, of morally-or-motivationally-questionable intent.

Hilariously, Python `nose` is deprecated for `nose2`, so maybe you do add the `2` from the get go, to avoid a name change later.

'Snitch' moves away from Catchy reference. This is good or bad depending on the library aims. There's a long tradition of tribute bands selecting clever punning names. But can a tribute band play their own songs? Can a Funk band play Rock Music? Following Catch is a useful on-ramp for those familiar with its API. It could also over-constrain, say to stick with it's macro system when a some macro-free interface may be more appropriate.

Your stated aim in this regard
>* Limited subset of the [Catch2](https://github.com/catchorg/_Catch2_) API, see [Comparison with Catch2](https://github.com/cschreib/snatch#detailed-comparison-with-catch2).
>* Additional API not in Catch2, or different from Catch2:

I guess in the end direction will be decided by taste, user requests and desire to remain as minimal as appropriate. Whatever, it's good to have some wiggle room to change, evolve and pivot.

'Snitch' does not 'pass-off' as Catch; that seems right. Phil is less likely to send in the lawyers (he'd be more likely to take you down the pub, have a word, and bring in Kevlin Henney to bore you with shaggy-dog stories of yore)

**Reply by cschreib** [reply]
I just merged a PR renaming the library to _snitch_. Thanks for raising this issue! _snitch_ doesn't sound as nice, but it's not terrible, and that's the bar I was aiming for :D

> 'Snitch' is also semantically related to 'nose', a name already related to test in a Python test automation framework. Sniffing out bugs and errors. The logo could then be something nose-related in order to reinforce this interpretation over the murky image of an informant, of morally-or-motivationally-questionable intent.

That's a nice idea! I'll keep it in mind if ever we need a logo ;) 

> Following Catch is a useful on-ramp for those familiar with its API. It could also over-constrain, say to stick with it's macro system when a some macro-free interface may be more appropriate. [...] I guess in the end direction will be decided by taste, user requests and desire to remain as minimal as appropriate. Whatever, it's good to have some wiggle room to change, evolve and pivot.

Yes, I agree. The point of _snitch_ is to be its own standalone library, not a tribute. I chose to reproduce the _Catch2_ API because it is well used, and this can simplify migration. But this doesn't mean the library needs to be 100% like _Catch2_; it can change the bad parts they are stuck with, and add new stuff they don't have.

I'd love for _snitch_ to be macro-free, but not at the expense of usability/readability. For example, _Boost UT_ went 100% macro-free, and they did a great job at it, but still they had to publicly expose weird operator overloading to support expression decomposition: `expect(that % i == 2)`. They need the `%` there because as per today's C++ rules, you need an operator to start the decomposition. We also use one, but it's hidden inside the `CHECK()` macro. Probably we could get rid of this when C++ supports reflection, but that's not there yet. Until then, macros still provide the nicest API.

**Comment by willwray** [comment]
Or, doubling-down on Catchiness:

'Catch20' the C++20 simplified Catch2 (it'll get old quick, like 11-labelled libraries)

'Catch22` the obvious one

'Catch2022' combines the above.

 ...only ten days left to make this name-change, before we roll on to another year, another C++ revision, another thing...

**Comment by willwray** [comment]
Congratulations on v1.0 :tada:

Link to the [reddit](https://www.reddit.com/r/cpp/comments/10dac2x/snitch_v10_a_lightweight_c20_testing_framework/) announcement.

It might be occasion for an 'announcement' category Discussion here too
(and that might help get Discussions rolling into new topic threads)
(instead of tagging on the end of this Welcome thread...).

I'll update and get testing your testing framework.

Onwards and upwards!

**Reply by cschreib** [reply]
Cheers! And good idea, I have just created the announcement in this forum. Feel free to open a new thread if you have more feedback, there are dedicated discussions "rooms" (how are they called nowadays?) for that.

**Comment by tocic** [comment]
IDK where to post this, but want to share that _snitch_ is mentioned in [the Stack Overflow 2023 developer survey](https://stackoverflow.blog/2023/05/08/the-2023-developer-survey-is-now-live/), congrats!

**Comment by cschreib** [comment]
I took the survey last night and noticed as well :) So random!
